src/effcheck.ml

```diff
@@ -34,7 +34,7 @@ let tcheck_lit l =
   | LitStr _ -> (String, no_eff)
 ;;
 
-let check_opt_invariants (typ, name, payload_lst) =
+let check_option_invars variant_constr =
```

Comment:
You named this parameter `variant_constr` but it is really a `term`; there is a mismatch in the name here. I'm not convinced that it's the right choice to pass the whole term here; especially if you filter on the type in the main function (as I suggest in my other comment), then by the time you call this function you already know that the type is `Option ty`, so you should be able to refine the datatype passed to this function (maybe just the `ty` argument of `Option ty`, the variant name and the arguments).